go/store/types/value_store.go

```diff
@@ -322,59 +334,162 @@ func (lvs *ValueStore) WriteValue(ctx context.Context, v Value) (Ref, error) {
 
 func (lvs *ValueStore) Root(ctx context.Context) (hash.Hash, error) {
 	root, err := lvs.cs.Root(ctx)
-
 	if err != nil {
 		return hash.Hash{}, err
 	}
-
 	return root, nil
 }
 
 func (lvs *ValueStore) Rebase(ctx context.Context) error {
 	return lvs.cs.Rebase(ctx)
 }
 
-// Call with lvs.bufferMu locked. Blocks until doingGC == false, releasing the
-// lock while we are blocked. Returns with the lock held, doingGC == false.
-func (lvs *ValueStore) waitForGC() {
-	for lvs.doingGC {
+// Call with lvs.gcMu locked. Blocks until gcState == gcState_NoGC,
+// releasing the lock while blocked.
+//
+// Returns with the lock held and gcState == gcState_NoGC.
+func (lvs *ValueStore) waitForNoGC() {
+	for lvs.gcState != gcState_NoGC {
 		lvs.gcCond.Wait()
 	}
 }
 
-// Call without lvs.bufferMu held. If val == false, then doingGC must equal
-// true when this is called. We will set it to false and return without
-// lvs.bufferMu held. If val == true, we will set doingGC to true and return
-// with lvs.bufferMu not held.
+// Call without lvs.gcMu locked.
 //
-// When val == true, this routine will block until it has a unique opportunity
-// to toggle doingGC from false to true while holding the lock.
-func (lvs *ValueStore) toggleGC(val bool) {
+// Will block the caller until gcState != gcState_Finalizing.
+//
+// When this function returns, ValueStore is guaranteed to be in a state such
+// that lvs.gcAddChunk will not return ErrAddChunkMustBlock. This function
+// returns a finalizer which must be called. gcAddChunk will not return
+// MustBlock until after the finalizer which this function returns is called.
+//
+// The critical sections delimited by `waitForNotFinalizingGC` and its return
+// value should fully enclose any critical section which takes `bufferMu`.
+//
+// These sections are coping with the fact that no call into NomsBlockStore
+// while we hold `lvs.bufferMu` is allowed to see `ErrAddChunkMustBlock` or we
+// will block with the lock held. While the lock is held, reads cannot
+// progress, and the GC process will not complete.
+
+func (lvs *ValueStore) waitForNotFinalizingGC() func() {
```

Comment:
nit: could this be refactored to a take a callback rather than returning a finalizer?